src/jv_type_array.c

```diff
@@ -0,0 +1,282 @@
+
+  if (*pstart < 0) *pstart = 0;
+  if (*pstart > len) *pstart = len;
+  if (*pend > len) *pend = len;
+  if (*pend < *pstart) *pend = *pstart;
+}
+
+
+int jvp_array_contains(jv a, jv b) {
+  int r = 1;
+  jv_array_foreach(b, bi, belem) {
+    int ri = 0;
+    jv_array_foreach(a, ai, aelem) {
+      if (jv_contains(aelem, jv_copy(belem))) {
+        ri = 1;
+        break;
+      }
+    }
+    jv_free(belem);
+    if (!ri) {
+      r = 0;
+      break;
+    }
+  }
+  return r;
+}
+
+
+/*
+ * Public
+ */
+
+static jv jvp_array_slice(jv a, int start, int end) {
+  assert(JVP_HAS_KIND(a, JV_KIND_ARRAY));
+  int len = jvp_array_length(a);
+  jvp_clamp_slice_params(len, &start, &end);
+  assert(0 <= start && start <= end && end <= len);
+
+  // FIXME: maybe slice should reallocate if the slice is small enough
+  if (start == end) {
+    jv_free(a);
+    return jv_array();
+  }
+
+  if (a.offset + start >= 1 << (sizeof(a.offset) * CHAR_BIT)) {
+    jv r = jv_array_sized(end - start);
+    for (int i = start; i < end; i++)
+      r = jv_array_append(r, jv_array_get(jv_copy(a), i));
+    jv_free(a);
+    return r;
+  } else {
+    a.offset += start;
+    a.size = end - start;
+    return a;
+  }
+}
+
+/*
+ * Arrays (public interface)
+ */
+
+jv jv_array_sized(int n) {
+  return jvp_array_new(n);
+}
+
+jv jv_array() {
+  return jv_array_sized(16);
+}
+
+int jv_array_length(jv j) {
+  assert(JVP_HAS_KIND(j, JV_KIND_ARRAY));
+  int len = jvp_array_length(j);
+  jv_free(j);
+  return len;
+}
+
+jv jv_array_get(jv j, int idx) {
+  assert(JVP_HAS_KIND(j, JV_KIND_ARRAY));
+  jv* slot = jvp_array_read(j, idx);
+  jv val;
+  if (slot) {
+    val = jv_copy(*slot);
+  } else {
+    val = jv_invalid();
+  }
+  jv_free(j);
+  return val;
+}
+
+jv jv_array_set(jv j, int idx, jv val) {
+  assert(JVP_HAS_KIND(j, JV_KIND_ARRAY));
+
+  if (idx < 0)
+    idx = jvp_array_length(j) + idx;
+  if (idx < 0) {
+    jv_free(j);
+    jv_free(val);
+    return jv_invalid_with_msg(jv_string("Out of bounds negative array index"));
+  }
+  // copy/free of val,j coalesced
+  jv* slot = jvp_array_write(&j, idx);
+  jv_free(*slot);
+  *slot = val;
+  return j;
+}
+
+jv jv_array_append(jv j, jv val) {
+  // copy/free of val,j coalesced
+  return jv_array_set(j, jv_array_length(jv_copy(j)), val);
+}
+
+jv jv_array_concat(jv a, jv b) {
+  assert(JVP_HAS_KIND(a, JV_KIND_ARRAY));
+  assert(JVP_HAS_KIND(b, JV_KIND_ARRAY));
+
+  // FIXME: could be faster
+  jv_array_foreach(b, i, elem) {
+    a = jv_array_append(a, elem);
+  }
+  jv_free(b);
+  return a;
+}
+
+jv jv_array_slice(jv a, int start, int end) {
+  assert(JVP_HAS_KIND(a, JV_KIND_ARRAY));
+  // copy/free of a coalesced
+  return jvp_array_slice(a, start, end);
+}
+
+jv jv_array_indexes(jv a, jv b) {
+  jv res = jv_array();
+  int idx = -1;
+  jv_array_foreach(a, ai, aelem) {
+    jv_free(aelem);
+    jv_array_foreach(b, bi, belem) {
+      if (!jv_equal(jv_array_get(jv_copy(a), ai + bi), jv_copy(belem)))
```

Comment:
remove the `jv_copy` from `belem` here, and you can skip the `jv_free(belem)` below.

Reply:
👍